### PR TITLE
CLOUDP-112092: Support decryption of Log Encryption Key using KMIP

### PR DESCRIPTION
**Pull request by tibulca**
## Proposed changes

_Jira ticket:_ CLOUDP-112092

Closes #[issue number]

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongocli/blob/master/CONTRIBUTING.md) (if appropriate)
- [x] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](https://github.com/mongodb/mongocli/blob/master/e2e/E2E-TESTS.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: internal/cli/opsmanager/logs/list_key_provider.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ type KeyProviderListOpts struct {
 	fs   afero.Fs
 }
 
-var listTmpl = `{{ range . }}{{ .Provider }}:{{if .Filename}} Filename = {{ .Filename }}{{end}}{{if .UniqueKeyID}} Unique Key ID = "{{ .UniqueKeyID }}"{{end}}{{if .KMIPServerName}} KMIP Server Name = "{{ .KMIPServerName }}"{{end}}{{if .KMIPPort}} KMIP Port = "{{ .KMIPPort }}"{{end}}{{if .KeyWrapMethod}} Key Wrap Method = "{{ .KeyWrapMethod }}"{{end}}
+var listTmpl = `{{ range . }}{{ .Provider }}:{{if .Filename}} Filename = {{ .Filename }}{{end}}{{if .UID}} Unique Key ID = "{{ .UID }}"{{end}}{{if .KMIPServerName}} KMIP Server Name = "{{index .KMIPServerName 0}}"{{end}}{{if .KMIPPort}} KMIP Port = "{{ .KMIPPort }}"{{end}}{{if .KeyWrapMethod}} Key Wrap Method = "{{ .KeyWrapMethod }}"{{end}}
```
Comment (tibulca) [inline internal/cli/opsmanager/logs/list_key_provider.go:33]: checking why the header format changed with server team

File: internal/decryption/encrypted_audit_log.go [hunk 2]
```diff
@@ -30,9 +30,9 @@ type AuditLogLineKeyStoreIdentifier struct {
 	// localKey
 	Filename *string `json:"filename,omitempty"`
 	// kmip
-	UniqueKeyID    *string                         `json:"uniqueKeyID,omitempty"`
-	KMIPServerName *string                         `json:"kmipServerName,omitempty"`
-	KMIPPort       *string                         `json:"kmipPort,omitempty"`
+	UID            *string                         `json:"uniqueKeyID,omitempty"`
+	KMIPServerName []string                        `json:"kmipServerName,omitempty"`
+	KMIPPort       *int                            `json:"kmipPort,omitempty"`
```
Comment (tibulca) [inline internal/decryption/encrypted_audit_log.go:35]: checking why the header format changed with server team

**Comment by tibulca** [comment]
e2e tests are implemented here https://jira.mongodb.org/browse/CLOUDP-116175 (WIP)

File: internal/decryption/kmip/client.go [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+	"fmt"
+
+	"crypto/tls"
+	"crypto/x509"
+
+	kmip "github.com/gemalto/kmip-go"
+	"github.com/gemalto/kmip-go/kmip14"
+	"github.com/gemalto/kmip-go/kmip20"
+	"github.com/gemalto/kmip-go/ttlv"
+	"github.com/pkg/errors"
```
Comment (andreaangiolillo) [inline internal/decryption/kmip/client.go:28]: remove the new lines here and use go imports

**Review by andreaangiolillo — COMMENTED** [review]
just one comment. 👍

File: internal/decryption/kmip/client.go (hunk 3 above)
Comment (tibulca) [reply internal/decryption/kmip/client.go]: done, thx

**Review by andreaangiolillo — DISMISSED** [review]
LGTM

File: internal/cli/opsmanager/logs/list_key_provider.go (hunk 1 above)
Comment (tibulca) [reply internal/cli/opsmanager/logs/list_key_provider.go]: for now, there won't be any other header change. Adapted the code to support multiple server names:
https://jira.mongodb.org/browse/CLOUDP-116151?focusedCommentId=4431895&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-4431895

File: internal/decryption/keyproviders/kmip.go [hunk 4]
```diff
@@ -48,20 +49,46 @@ type KMIPEncryptedKey struct {
...
 	if err != nil {
 		return nil, err
 	}
 
-	kmipClient, err := ki.kmipClient()
-	if err != nil {
-		return nil, err
+	var clientError error
+	collectErr := func(err error, serverName string) {
+		if clientError == nil {
```
Comment (andreaangiolillo) [inline internal/decryption/keyproviders/kmip.go:63]: [q] in which scenario, clientError would be different than nil? where are you setting this value?

**Review by andreaangiolillo — DISMISSED** [review]

File: go.sum [hunk 5]
```diff
@@ -68,6 +68,9 @@ github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuy
 github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
 github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf/go.mod h1:ybxpYRFXyAe+OPACYpWeL0wqObRcbAqCMya13uyzqw0=
 github.com/alecthomas/units v0.0.0-20190717042225-c3de453c63f4/go.mod h1:ybxpYRFXyAe+OPACYpWeL0wqObRcbAqCMya13uyzqw0=
+github.com/ansel1/merry v1.5.0/go.mod h1:wUy/yW0JX0ix9GYvUbciq+bi3jW/vlKPlbpI7qdZpOw=
```
Comment (gssbzn) [inline go.sum:71]: can you please do a `go mod tidy`? it feels like there are some lefover heres from package updates

File: internal/decryption/keyproviders/kmip.go [hunk 6]
```diff
@@ -14,7 +14,14 @@
 
 package keyproviders
 
-import "fmt"
+import (
+	"os"
+
+	"github.com/mongodb/mongocli/internal/decryption/aes"
+	"github.com/mongodb/mongocli/internal/decryption/kmip"
+	"github.com/pkg/errors"
```
Comment (gssbzn) [inline internal/decryption/keyproviders/kmip.go:22]: please don't use, this package is archived and unmaintained

**Review by gssbzn — CHANGES_REQUESTED** [review]
seems like you are using https://github.com/pkg/errors, that package was abandoned once go1.13 errors became official, please let's replace that lib with the standard lib

File: internal/decryption/keyproviders/kmip.go [hunk 7]
```diff
@@ -23,16 +30,124 @@ const (
...
+	if err != nil {
+		return nil, err
+	}
+
+	var clientError error
+	collectErr := func(err error, serverName string) {
+		if clientError == nil {
+			clientError = err
+		} else {
+			clientError = errors.Wrapf(err, "'%s': %s", serverName, err.Error())
```
[the remark this answers is not in the document]
Comment (tibulca) [reply internal/decryption/keyproviders/kmip.go:66]: done

File: internal/decryption/keyproviders/kmip.go (hunk 4 above)
Comment (tibulca) [reply internal/decryption/keyproviders/kmip.go:63]: in the scenario where we have multiple KMIP servers, if the first one is failing then clientError will be initialized at this line. For the next servers that are failing will, the error will be concatenated with the previous clientError (that's why the name of the function, collectErr).

File: go.sum (hunk 5 above)
Comment (tibulca) [reply go.sum:71]: did that but this package is still there, it's indirect dependency of github.com/gemalto/kmip-go...

File: internal/decryption/kmip/client.go [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+	}
+
+	rootCAs := x509.NewCertPool()
+	if !rootCAs.AppendCertsFromPEM(config.RootCertificate) {
+		return nil, errors.New("failed to load server's root certificate")
+	}
+
+	certificate, err := tls.X509KeyPair(config.ClientCertificate, config.ClientPrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load client key and certificate: %s", err.Error())
```
Comment (gssbzn) [inline internal/decryption/kmip/client.go:148]: [non blocking/separate PR] you can do 
```suggestion
		return nil, fmt.Errorf("failed to load client key and certificate: %w", err)
```

**Review by gssbzn — APPROVED** [review]
LGTM thanks for removing the deprecated package

File: internal/decryption/kmip/client.go (hunk 8 above)
Comment (tibulca) [reply internal/decryption/kmip/client.go:148]: true :), will include it in my next PR (e2e for this code)

File: internal/decryption/kmip/client.go [hunk 9]
```diff
@@ -0,0 +1,385 @@
+// Copyright 2022 MongoDB Inc
```
Comment (fmenezes) [inline internal/decryption/kmip/client.go:1]: not blocking but this seems a lot of code that is not tested, would you be able to send some unit tests (perhaps on a follow-up pr)?

Comment (tibulca) [reply internal/decryption/kmip/client.go:1]: doing this as part of https://jira.mongodb.org/browse/CLOUDP-116175 E2E smoke tests - KMIP

**Pull request merged**